src/ott/neural/models/base_solver.py

```diff
@@ -0,0 +1,485 @@
+    indices_source = indices // tmat.shape[1]
+    indices_target = indices % tmat.shape[1]
+    return tuple(
+        b[indices_source, :] if b is not None else None for b in source_arrays
+    ), tuple(
+        b[indices_target, :] if b is not None else None for b in target_arrays
+    )
+
+  def _sample_conditional_indices_from_tmap(
+      self,
+      key: jax.random.PRNGKeyArray,
+      tmat: jnp.ndarray,
+      k_samples_per_x: Union[int, jnp.ndarray],
+      source_arrays: Tuple[jnp.ndarray, ...],
+      target_arrays: Tuple[jnp.ndarray, ...],
+      *,
+      source_is_balanced: bool,
+  ) -> Tuple[jnp.array, jnp.array]:
+    batch_size = tmat.shape[0]
+    left_marginals = tmat.sum(axis=1)
+    if not source_is_balanced:
+      key, key2 = jax.random.split(key, 2)
+      indices = jax.random.choice(
+          key=key2,
+          a=jnp.arange(len(left_marginals)),
+          p=left_marginals,
+          shape=(len(left_marginals),)
+      )
+    else:
+      indices = jnp.arange(batch_size)
+    tmat_adapted = tmat[indices]
+    indices_per_row = jax.vmap(
+        lambda tmat_adapted: jax.random.choice(
+            key=key,
+            a=jnp.arange(batch_size),
+            p=tmat_adapted,
+            shape=(k_samples_per_x,)
+        ),
+        in_axes=0,
+        out_axes=0,
+    )(
+        tmat_adapted
+    )
+
+    indices_source = jnp.repeat(indices, k_samples_per_x)
+    indices_target = jnp.reshape(
+        indices_per_row % tmat.shape[1], (batch_size * k_samples_per_x,)
+    )
+    return tuple(
+        jnp.reshape(b[indices_source], (k_samples_per_x, batch_size,
+                                        -1)) if b is not None else None
+        for b in source_arrays
+    ), tuple(
+        jnp.reshape(b[indices_target, :], (k_samples_per_x, batch_size,
+                                           -1)) if b is not None else None
+        for b in target_arrays
+    )
+
+  def _get_sinkhorn_match_fn(
+      self,
+      ot_solver: Any,
+      epsilon: float = 1e-2,
+      cost_fn: Optional[costs.CostFn] = None,
+      scale_cost: Union[bool, int, float,
+                        Literal["mean", "max_norm", "max_bound", "max_cost",
+                                "median"]] = "mean",
+      tau_a: float = 1.0,
+      tau_b: float = 1.0,
+      *,
+      filter_input: bool = False,
+  ) -> Callable:
+
+    @jax.jit
+    def match_pairs(
+        x: jnp.ndarray, y: jnp.ndarray
+    ) -> Tuple[jnp.ndarray, jnp.ndarray, jnp.ndarray, jnp.ndarray]:
+      geom = pointcloud.PointCloud(
+          x, y, epsilon=epsilon, scale_cost=scale_cost, cost_fn=cost_fn
+      )
+      return ot_solver(
+          linear_problem.LinearProblem(geom, tau_a=tau_a, tau_b=tau_b)
+      ).matrix
+
+    @jax.jit
+    def match_pairs_filtered(
+        x_lin: jnp.ndarray, x_quad: jnp.ndarray, y_lin: jnp.ndarray,
+        y_quad: jnp.ndarray
+    ) -> Tuple[jnp.ndarray, jnp.ndarray, jnp.ndarray, jnp.ndarray]:
```

Comment:
Wrong return type; but overall I think the function `match_pairs_filtered` shouldn't exist.